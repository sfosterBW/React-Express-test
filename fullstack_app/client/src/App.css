.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #09d3ac;
}

#fruit-form {
  background-color: grey;
  padding: 5vmin 8vmin;
  border-radius: 0.5vmin;
}

.input-wrapper {
  display: flex;
  margin-bottom: 2vmin;
}

.input-wrapper label {
  display: inline-block;
  width: 20vmin;
  margin-right: 2vmin;
  font-weight: bold;
  text-align: right;
}

.input-wrapper input[type=text] {
  height: 4vmin;
  font-size: 4vmin;
}

.input-wrapper input[type=radio] {
  margin-left: 1vmin;
  margin-right: 4vmin;
}

#fruit-form button {
  display: flex;
  margin: 0 auto;
  padding: 2vmin 4vmin;
  font-size: 2vmin;
  border-radius: 0.5vmin;
}

#fruit-wrapper {

}

#fruit-wrapper table, th, td {
  border: 1px solid black;
  min-width: 20vmin;
  background-color: white;
  color: black;
  margin: 4vmin;
}

#fruit-wrapper table tr td p{
  margin: 2vmin;
}

#fruit-wrapper .checkbox {
  margin-right: 12px;
}
