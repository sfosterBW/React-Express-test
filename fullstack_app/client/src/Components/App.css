.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #09d3ac;
}

#fruit-form {
  background-color: grey;
  padding: 5vmin 8vmin;
  border-radius: 0.5vmin;
}

#fruit-form .input-wrapper {
  display: flex;
  margin-bottom: 2vmin;
}

#fruit-form .input-wrapper label {
  display: inline-block;
  width: 20vmin;
  margin-right: 2vmin;
  font-weight: bold;
  text-align: right;
}

#fruit-form .input-wrapper input[type=text] {
  height: 4vmin;
  font-size: 3vmin;
}

#fruit-form .input-wrapper input[type=radio] {
  margin-left: 1vmin;
  margin-right: 4vmin;
}

#fruit-form button {
  display: flex;
  margin: 0 auto;
  padding: 2vmin 4vmin;
  font-size: 2vmin;
  border-radius: 0.5vmin;
}

#fruit-table {}

#fruit-table table {
  border-collapse: collapse;
}

#fruit-table thead {
  border-bottom: 1px solid black;
}

#fruit-table table, th, td {
  min-width: 20vmin;
  background-color: white;
  color: black;
  margin: 4vmin;
}

#fruit-table tbody tr {
  border-bottom: 1px solid lightgrey;
}

#fruit-table table tr td p {
  margin: 2vmin;
}

#fruit-table .checkbox {
  margin-right: 12px;
}
