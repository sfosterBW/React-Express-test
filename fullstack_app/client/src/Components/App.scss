//Global variables
$primary-background: #282c34;
$primary-text: white;
$primary-link-text: #09d3ac;
$primary-shadow: rgba(0,0,0,0.2);
$secondary-background: white;
$secondary-text: black;
$secondary-dark-border: black;
$secondary-light-border: lightgrey;

.App {
  text-align: center;

  .App-header {
    background-color: $primary-background;
    min-height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    font-size: calc(10px + 2vmin);
    color: $primary-text;

    form {
      background: grey;
    }
  }
}

.App-link {
  color: $primary-link-text;
}

.fruit-form {
  padding: 5vmin 8vmin;
  border-radius: 0.5vmin;

  button {
    display: flex;
    margin: 0 auto;
    padding: 2vmin 4vmin;
    font-size: 2vmin;
    border-radius: 0.5vmin;
  }
}

.input-wrapper {
  display: flex;
  margin-bottom: 2vmin;

  label {
    display: inline-block;
    width: 20vmin;
    margin-right: 2vmin;
    font-weight: bold;
    text-align: right;
  }

  input[type=text] {
    height: 4vmin;
    font-size: 3vmin;
  }

  input[type=radio] {
    margin-left: 1vmin;
    margin-right: 4vmin;
  }
}

table {
  border-collapse: collapse;
  background-color: $secondary-background;
  color: black;
  margin: 4vmin;

  thead {
    border-bottom: 1px solid $secondary-dark-border;
  }

  th, td {
    min-width: 20vmin;
    margin: 4vmin;
    padding: 2vmin;
  }

  tbody tr {
    border-bottom: 1px solid $secondary-light-border;
  }

  input[type="checkbox"] {
    margin-right: 12px;
  }

}

.modal-wrapper {
  position: absolute;
  bottom: 0;
  right: 0;
  left: 0;
  top: 0;
  background: $primary-shadow;
  display: flex;
  justify-content: center;
  align-items: center;

  .modal {
    position: relative;
    border-radius: 5px;
    background: $secondary-background;
    color: $secondary-text;
  }

  .closeButton {
    position: absolute;
    top: 1.5vmin;
    right: 0px;
    transform: translateY(-50%);
    cursor: pointer;
  }
}
