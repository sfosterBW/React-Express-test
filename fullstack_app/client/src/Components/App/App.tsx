import React, { FC, useEffect, useCallback } from 'react'
import { Router } from "@reach/router"
import { useSelector, useDispatch } from 'react-redux'
import { RootState } from '../../utils/store'
import fruitService from '../../utils/api'
import { getFruits } from '../../utils/actions'
import Modal from '../Modal/Modal'
import Nav from '../Nav/Nav'

import AddFruit from '../../Pages/AddFruit'
import EditFruit from '../../Pages/EditFruit'
import Home from '../../Pages/Home'
import styles from './App.module.css'

const App: FC = () => {

  const selectFruits = (state: RootState) => state.fruit.data
  const fruits = useSelector(selectFruits)
  const selectModalFruit = (state: RootState) => state.modal.fruit
  const modalFruit = useSelector(selectModalFruit)
  const dispatch = useDispatch()
  //Used to ensure the same values are used on re-render
  const effectDispatch = useCallback(dispatch, [])

  useEffect(() => {
    const initFruit = () => {
      fruitService.fetchFruit()
        .then(res => effectDispatch(getFruits(res)))
    }
    initFruit()
  }, [effectDispatch])

  return (
    <div className="App">
      <Nav />
      <main className={styles.main}>
        <Router className={styles.main}>
          <Home path="/" title="Add Fruit and stuff" />
          <AddFruit path="/add" title="Add Fruit and stuff" />
          <EditFruit fruits={fruits} path="/edit" title="True table"/>
        </Router>
      </main>
      <Modal fruit={modalFruit} />
    </div>
  )
}

export default App
