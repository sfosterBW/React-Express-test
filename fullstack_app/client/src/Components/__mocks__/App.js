const testData = {
  falseCase: {
    _id: 1,
    name: "False Case",
    best: false
  },
  trueCase: {
    _id: 2,
    name: "True Case",
    best: true
  }
}

export default function getFruitList() {
  return testData.falseCase
}
