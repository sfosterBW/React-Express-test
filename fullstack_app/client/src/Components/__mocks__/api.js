'use strict'
module.exports = {
  fetchFruit: () => {
    return Promise.resolve({
      status: 200,
      data: [
        {
          _id: 1,
          name: "False Case",
          best: false
        },
        {
          _id: 2,
          name: "True Case",
          best: true
        }
      ]
    })
  }
}
