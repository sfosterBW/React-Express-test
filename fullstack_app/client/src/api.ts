import axios from 'axios'
import { Fruit, NewFruit } from './Components/interfaces'

function handleError(error: any) {
  console.error(error)
}

export async function getFruitList() {
  try {
    const response = await axios.get("/fruit-api/list")
    const fruitList: Fruit[] = response.data
    const result = {status: response.status, data: fruitList}
    return result
  } catch (error) {
    const emptyList: Fruit[] = []
    handleError(error)
    const result = {status: 500, data: emptyList}
    return result
  }
}

export async function createFruit(fruit: NewFruit) {
    try {
      const response = await axios.post("/fruit-api/new", { new: fruit })
      return response.status
    }
    catch (error) {
      handleError(error)
      return 500
    }
}

export async function updateFruit(fruit: Fruit) {
  try {
    const response = await axios.put("/fruit-api/update", fruit)
    return response.status
  }
  catch (error) {
    handleError(error)
    return 500
  }
}

export async function deleteFruit(id: number) {
  const deleteData = { params: { id: id } }
  try {
    const response = await axios.delete("/fruit-api/delete", deleteData)
    return response.status
  }
  catch (error) {
    handleError(error)
    return 500
  }
}
