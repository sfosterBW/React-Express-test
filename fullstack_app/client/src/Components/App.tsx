import React, { FC, useState, useEffect } from 'react'
import { createFruit, deleteFruit, fetchFruit, updateFruit } from '../utils/api'
import { IFruit } from '../utils/interfaces'
import { toggleAlert } from '../utils/actions'
import { useDispatch } from 'react-redux'
import './App.scss'
import Alert from './Alert'
import Form from './Form'
import Modal from './Modal'
import Row from './Row'
import Table from './Table'

interface Props {

}

const App: FC<Props> = () => {

  const [fruitList, setFruitList] = useState<Array<IFruit>>([])
  const [modalToggle, setModalToggle] = useState<boolean>(false)
  const [modalFruit, setModalFruit] = useState<IFruit>()
  const dispatch = useDispatch()

  useEffect(() => {
    async function initList() {
      const res = await fetchFruit()
      res ? setFruitList(res.data) : setFruitList([])
    }
    initList()
  }, [])

  const handleBestChange = async (fruit: IFruit) => {
    const updatedFruit = fruit
    updatedFruit.best = !updatedFruit.best
    const res = await updateFruit(updatedFruit)
    res.status === 200 ? setFruitList(res.data) : handleError("handleBestChange")
  }

  const handleError = (error: any) => {
    console.error("App", error)
    dispatch(toggleAlert(true))
  }

  const handleSubmit = async (fruit: IFruit) => {
    if (fruit._id < 0) {
      const res = await createFruit(fruit)
      res.status === 200 ? setFruitList(res.data) : handleError("handleSubmit add")
    } else if (fruit._id >= 0) {
      const res = await updateFruit(fruit)
      res.status === 200 ? setFruitList(res.data) : handleError("handleSubmit update")
    } else {
      console.log("handle submit unexpected error")
    }
    setModalToggle(false)
  }

  const handleRemove = async (fruit: IFruit) => {
    const res = await deleteFruit(fruit._id)
    res.status === 200 ? setFruitList(res.data) : handleError("handleSubmit")
  }

  const displayFruitList = (best: boolean) => {
    const fruitRows = fruitList
      .filter((i: IFruit) => i.best === best)
      .map((i: IFruit) =>
        <Row
          key={`${i._id}${i.name}`}
          fruit={i}
          handleRemove={() => { handleRemove(i) }}
          handleEdit={() => { handleBestChange(i) }}
          openModal={() => {
            setModalToggle(true)
            setModalFruit(i)
          }} />)
    const tableBody = <tbody>{fruitRows}</tbody>
    return tableBody
  }

  const displayModal = () => {
    if (modalFruit) {
      return (
        <Modal
          form={<Form handleSubmit={handleSubmit} fruit={modalFruit} />}
          onClose={() => { setModalToggle(false) }}
          title="This is a modal" />
      )
    } else {
      console.log("No modal fruit defined yet")
      return null
    }
  }

  return (
    <div className="App">
      <header className="App-header">
        <h1>Fruit dashboard</h1>
        <Alert />
        <Form handleSubmit={handleSubmit} />
        <Table rows={displayFruitList(true)} title="True table" />
        <Table rows={displayFruitList(false)} title="False table" />
      </header>
      {modalToggle && displayModal()}
    </div>
  )
}

export default App
