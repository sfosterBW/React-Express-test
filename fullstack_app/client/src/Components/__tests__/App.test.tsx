import React from 'react'
import { mount} from 'enzyme'
import renderer from 'react-test-renderer'
import * as api from './../../utils/api'

import App from '../App'



describe('the app component', () => {

  jest.mock('./../../utils/api')
  const getSpy = jest.spyOn(api, 'fetchFruit')
  const appComponent = <App />
  const app = mount(appComponent)


  it('renders with the correct structure', () => {
    expect(app).toBeDefined()
    expect(app.find('div.App')).toHaveLength(1)
    expect(app.find('header.App-header')).toHaveLength(1)
    expect(app.find('Form')).toHaveLength(1)
    expect(app.find('Table')).toHaveLength(2)
    expect(app.find('Table').at(0).props().title).toEqual("True table")
    expect(app.find('Table').at(1).props().title).toEqual("False table")
  })

  it('functions as expected', () => {
    expect(getSpy).toHaveBeenCalled()
  })

  it('renders the same as last time', () => {
    const tree = renderer
      .create(appComponent)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
