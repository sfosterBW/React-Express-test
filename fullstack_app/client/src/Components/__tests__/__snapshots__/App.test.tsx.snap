// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the app component renders the same as last time 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <h1>
      Fruit dashboard
    </h1>
    <div />
    <form
      className="fruit-form"
      onSubmit={[Function]}
    >
      <h2>
        List out your favourite fruit
      </h2>
      <div
        className="input-wrapper"
      >
        <label
          htmlFor="name"
        >
          Add a fruit:
        </label>
        <input
          name="name"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="input-wrapper"
      >
        <label
          htmlFor="best"
        >
          Is it best?:
        </label>
        <input
          checked={false}
          name="best"
          onChange={[Function]}
          type="checkbox"
          value="false"
        />
      </div>
      <button>
        Add new fruit
      </button>
    </form>
    <div
      className="fruit-table"
    >
      <h2>
        True table
      </h2>
      <table>
        <thead>
          <tr>
            <th>
              ID
            </th>
            <th>
              Fruit
            </th>
            <th>
              Best?
            </th>
            <th>
              Edit
            </th>
            <th>
              Update
            </th>
            <th>
              Remove
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
    <div
      className="fruit-table"
    >
      <h2>
        False table
      </h2>
      <table>
        <thead>
          <tr>
            <th>
              ID
            </th>
            <th>
              Fruit
            </th>
            <th>
              Best?
            </th>
            <th>
              Edit
            </th>
            <th>
              Update
            </th>
            <th>
              Remove
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </header>
</div>
`;
